cc/attestation/cose.cc

```diff
@@ -0,0 +1,127 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cc/attestation/cose.h"
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+#include "libcppbor/include/cppbor/cppbor.h"
+#include "libcppbor/include/cppbor/cppbor_parse.h"
+
+namespace oak::attestation {
+
+absl::StatusOr<CoseSign1> CoseSign1::Deserialize(const std::vector<uint8_t>& data) {
+  auto [item, end, error] = cppbor::parse(data);
+  if (!error.empty()) {
+    return absl::InvalidArgumentError(absl::StrCat("couldn't parse COSE Sign1: ", error));
+  }
+  if (item->type() != cppbor::ARRAY) {
+    return UnexpectedCborTypeError("COSE Sign1", cppbor::ARRAY, item->type());
+  }
+  const cppbor::Array* array = item->asArray();
+  if (array->size() != 4) {
+    return absl::InvalidArgumentError(
+        absl::StrCat("invalid COSE Sign1 CBOR array size, expected 4, found ", array->size()));
+  }
+
+  const std::unique_ptr<cppbor::Item>& protected_headers = array->get(0);
+  if (protected_headers->type() != cppbor::BSTR) {
+    return UnexpectedCborTypeError("protected_headers", cppbor::BSTR, protected_headers->type());
+  }
+  const std::unique_ptr<cppbor::Item>& unprotected_headers = array->get(1);
+  if (unprotected_headers->type() != cppbor::MAP) {
+    return UnexpectedCborTypeError("unprotected_headers", cppbor::MAP, unprotected_headers->type());
+  }
+  const std::unique_ptr<cppbor::Item>& payload = array->get(2);
+  if (payload->type() != cppbor::BSTR) {
+    return UnexpectedCborTypeError("payload", cppbor::BSTR, payload->type());
+  }
+  const std::unique_ptr<cppbor::Item>& signature = array->get(3);
+  if (signature->type() != cppbor::BSTR) {
+    return UnexpectedCborTypeError("signature", cppbor::BSTR, signature->type());
+  }
+
+  return CoseSign1(protected_headers->asBstr(), unprotected_headers->asMap(), payload->asBstr(),
+                   signature->asBstr(), std::move(item));
+}
+
+absl::StatusOr<CoseKey> CoseKey::Deserialize(const std::vector<uint8_t>& data) {
+  auto [item, end, error] = cppbor::parse(data);
+  if (!error.empty()) {
+    return absl::InvalidArgumentError(absl::StrCat("couldn't parse COSE Key: ", error));
+  }
+  if (item->type() != cppbor::MAP) {
+    return UnexpectedCborTypeError("COSE Key", cppbor::MAP, item->type());
+  }
+  const cppbor::Map* map = item->asMap();
+  if (map->size() < 5) {
+    return absl::InvalidArgumentError(
+        absl::StrCat("invalid COSE Key CBOR map size, expected >= 5, found ", map->size()));
+  }
+
+  const std::unique_ptr<cppbor::Item>& kty = map->get<int, int>(KTY);
+  if (!kty) {
```

Comment:
[Internally](http://go/totw/141#two-kinds-of-null-pointer-checks) there is a slight preference towards the more verbose style: (kty == nullptr)

Reply:
Done